flask_security/core.py

```diff
@@ -461,30 +461,48 @@ class Security(object):
 
     :param app: The application.
     :param datastore: An instance of a user datastore.
+    :param register_blueprint: to register the Security blueprint or not.
+    :param login_form: set form for the login view
+    :param register_form: set form for the register view
+    :param confirm_register_form: set form for the confirm register view
+    :param forgot_password_form: set form for the forgot password view
+    :param reset_password_form: set form for the reset password view
+    :param change_password_form: set form for the change password view
+    :param send_confirmation_form: set form for the send confirmation view
+    :param passwordless_login_form: set form for the passwordless login view
+    :param anonymous_user: class to use for anonymous user
     """
 
     def __init__(self, app=None, datastore=None, **kwargs):
         self.app = app
         self.datastore = datastore
-
+        self.kwargs = dict(register_blueprint=True,
+                           login_form=None,
+                           register_form=None,
+                           confirm_register_form=None,
+                           forgot_password_form=None,
+                           reset_password_form=None,
+                           change_password_form=None,
+                           send_confirmation_form=None,
+                           passwordless_login_form=None,
+                           anonymous_user=None)
+        self.kwargs.update(kwargs)
+
+        self._state = None  # set by init_app
         if app is not None and datastore is not None:
-            self._state = self.init_app(app, datastore, **kwargs)
-
-    def init_app(self, app, datastore=None, register_blueprint=True,
-                 login_form=None, confirm_register_form=None,
-                 register_form=None, forgot_password_form=None,
-                 reset_password_form=None, change_password_form=None,
-                 send_confirmation_form=None, passwordless_login_form=None,
-                 anonymous_user=None):
+            self.init_app(app, datastore, **self.kwargs)
+
+    def init_app(self, app, datastore=None, **kwargs):
         """Initializes the Flask-Security extension for the specified
-        application and datastore implentation.
+        application and datastore implementation.
 
         :param app: The application.
         :param datastore: An instance of a user datastore.
-        :param register_blueprint: to register the Security blueprint or not.
         """
         self.app = app
-        self.datastore = datastore
+        if datastore is not None:
```

Comment:
IMHO we should not change the global datastore.

```python
datastore = datastore or self.datastore
kwargs.update(self.kwargs)

# ...

self._state  = state = _get_state(app, datastore, **kwargs)
```

Reply:
I made this change, but there could be one possibly surprising behaviour with the kwargs:
```
kwargs.update(self.kwargs)
```
Doing it this way, kwargs passed to __init__ would always take precedence over those passed to init_app. I'm not sure why somebody would want to pass an argument to the constructor and then later override it in the call to init_app, but, something to note.